core/src/Kokkos_Macros.hpp

```diff
@@ -166,6 +166,11 @@
 
 #if defined(KOKKOS_ENABLE_HIP)
 
+// FIXME_HIP Not defining NDEBUG makes some tests fail.
+#ifndef NDEBUG
+#define NDEBUG
+#endif
```

Comment:
Does that propagate to user code? I am not sure everybody would be happy with this.

Reply:
At the moment, this is needed for this pull request for all the tests to pass. IMHO it doesn't help if we go only half of the way and don't enforce it in user code just to break there code if they don't do it themselves.
Of course, it's a judgment call if this pull request rectifies messing with `NDEBUG` in user code.

Reply:
My problem is that we silently change the way `assert` works in user code. From a user perspective, it's going to be a headache to understand what's going on.

Reply:
This does propagate to user code. whats the actual problem solved by this? Is it maybe largely a register limit issue?